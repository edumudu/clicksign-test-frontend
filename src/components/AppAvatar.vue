<template>
  <img
    v-if="hasImage"
    :class="[$style.img]"
    :src="src"
    :alt="name"
    @error="onImageError"
  >

  <div
    v-else
    :class="[$style.img, $style.placeholder]"
    :style="{ backgroundColor: color }"
  >
    {{ firstLetter }}
  </div>
</template>

<script lang="ts">
import { defineComponent, computed, ref } from 'vue';

import { strToHexColor } from '@/utils/hexdecimal';

export default defineComponent({
  name: 'AppAvatar',

  props: {
    src: {
      type: String,
      default: '',
    },

    name: {
      type: String,
      default: '',
    },
  },

  setup(props) {
    const failedToLoad = ref(false);
    const hasImage = computed(() => !!props.src && !failedToLoad.value);
    const firstLetter = computed(() => props.name.charAt(0).toUpperCase());
    const color = computed(() => strToHexColor(props.name));

    const onImageError = () => {
      failedToLoad.value = true;
    };

    return {
      color,
      firstLetter,
      hasImage,

      onImageError,
    };
  },
});
</script>

<style lang="scss" module>
.img {
  $size: 24px;

  border-radius: 50%;
  width: $size;
  height: $size;
}

.placeholder {
  font-size: 1rem;
  display: grid;
  place-items: center;
  color: var(--white);
}
</style>
