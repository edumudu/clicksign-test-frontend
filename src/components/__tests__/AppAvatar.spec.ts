import { shallowMount, VueWrapper } from '@vue/test-utils';

import AppAvatar from '../AppAvatar.vue';

describe('AppAvatar.vue', () => {
  let wrapper: VueWrapper<any>;

  beforeEach(() => {
    wrapper = shallowMount(AppAvatar);
  });

  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('Should render first letter of the name when not have image', async () => {
    await wrapper.setProps({ name: 'John Doe' });

    expect(wrapper.text()).toBe('J');
  });

  it('Should render the avatar when exists', async () => {
    const src = 'https://picsum.photos/id/237/200/300';
    await wrapper.setProps({ src });

    expect(wrapper.element.nodeName).toBe('IMG');
    expect(wrapper.attributes('src')).toBe(src);
  });

  it('Should render the first letter of the name when image failed to load', async () => {
    const src = 'https://isnt-image.com';
    await wrapper.setProps({ src, name: 'Billie' });

    // Wait for the image to try to load
    setTimeout(() => expect(wrapper.text()).toBe('B'), 1);
  });
});
