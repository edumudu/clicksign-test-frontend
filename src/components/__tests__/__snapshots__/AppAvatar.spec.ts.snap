// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppAvatar.vue should match snapshot 1`] = `
<div
  class="img placeholder"
  style="background-color: rgb(0, 0, 0);"
/>
`;
